astropy/config/tests/test_configs.py

```diff
@@ -45,18 +73,78 @@ def test_paths():
         pytest.param("XDG_CONFIG_HOME", paths.get_config_dir_path, id="config"),
     ],
 )
+@pytest.mark.usefixtures("ignore_config_paths_global_state")
 def test_xdg_variables(monkeypatch, tmp_path, environment_variable, func):
     config_dir = tmp_path / "astropy"
     config_dir.mkdir()
     monkeypatch.setenv(environment_variable, str(tmp_path))
     assert func() == config_dir
 
 
-def test_set_temp_config(tmp_path, monkeypatch):
+ENV_VAR_AND_FUNC = [
+    pytest.param(
+        "XDG_CACHE_HOME",
+        paths.get_cache_dir_path,
+        id="xdg-cache",
+    ),
+    pytest.param(
+        "XDG_CONFIG_HOME",
+        paths.get_config_dir_path,
+        id="xdg-config",
+    ),
+]
+
+
+@pytest.mark.parametrize("env_var, func", ENV_VAR_AND_FUNC)
+@pytest.mark.usefixtures("ignore_config_paths_global_state")
+def test_env_variables_setup(monkeypatch, tmp_path, env_var, func):
+    default_path = func()
+    target_dir = tmp_path / "nonexistent"
+    monkeypatch.setenv(env_var, str(target_dir))
+
+    with pytest.warns(
+        UserWarning,
+        match=rf"^{env_var} is set to '",
+    ):
+        new_path = func()
+
+    assert new_path == default_path
+
+
+@pytest.mark.parametrize("env_var, func", ENV_VAR_AND_FUNC)
+@pytest.mark.usefixtures("ignore_config_paths_global_state")
+def test_env_variables_setup_writable_but_missing(monkeypatch, tmp_path, env_var, func):
+    # have the env var point to a writable location,
+    # where an 'astropy' subdir is missing, but can be created silently
+    target_dir = tmp_path
+    expected_path = tmp_path / "astropy"
+    monkeypatch.setenv(env_var, str(target_dir))
+
+    assert not expected_path.exists()
+    path = func()
+    assert expected_path.is_dir()
+    assert path == expected_path
```

Comment:
The failure for this check is also expected. This is because for this situation, `expected_path` is symlinked to `path`, but `path` is the actual value that is returned.

https://github.com/astropy/astropy/blob/822166bb1581cbe996fa5d15ade73a2613cccb48/astropy/config/paths.py#L178-L179

https://github.com/astropy/astropy/blob/822166bb1581cbe996fa5d15ade73a2613cccb48/astropy/config/paths.py#L210-L212